server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java

```diff
@@ -1311,7 +1311,10 @@ private SearchSourceBuilder parseXContent(XContentParser parser, boolean checkTr
                         SearchSourceBuilder.STORED_FIELDS_FIELD.getPreferredName(),
                         parser
                     );
-                    searchUsage.trackSectionUsage(STORED_FIELDS_FIELD.getPreferredName());
+                    if (storedFieldsContext.fetchFields() == false
+                        || (storedFieldsContext.fieldNames() != null && storedFieldsContext.fieldNames().size() > 0)) {
+                        searchUsage.trackSectionUsage(STORED_FIELDS_FIELD.getPreferredName());
+                    }
```

Comment:
Can you expand on the rationale behind this part of the change? I believ.e it changes the semantics on when we consider the section used, I'd like to better understand how and why

Reply:
Yeap, as mentioned in the description this was a bit outside of the scope of the NPE, but while I was looking at the original PR that introduced the `searchUsage` section tracking ([here](https://github.com/elastic/elasticsearch/pull/91528/files#r1023003540)) I noticed that the check for the existence of stored fields was only added if we were parsing an array ([code](https://github.com/elastic/elasticsearch/pull/91528/files#diff-8dbd67572112ad3b6add7c78a1c2b49f7bafbec32484bddbafdbb104000744d0R1357)) and not an object ([code](https://github.com/elastic/elasticsearch/pull/91528/files#diff-8dbd67572112ad3b6add7c78a1c2b49f7bafbec32484bddbafdbb104000744d0R1226)), but based on the discussion with Alan ([link](https://github.com/elastic/elasticsearch/pull/91528/files#r1023003540)) I was left with the impression that the latter was simply left out.  

I could very well have misunderstood something (hence why I also brought it up on the description). WDYT?  Should I remove this update from this PR?

Reply:
I am ok with this. I think I understand now. Would it make sense to factor out the conditional into a method that we call from both places?

Reply:
Yeah makes sense. The only minor "issue" that I see is that for all other cases the conditional is defined in-place even if duplicated (e.g. the checks for `searchUsage.trackSectionUsage(_SOURCE_FIELD.getPreferredName())`), so this would be a "first". 
Happy to do it, but maybe we could refactor similar cases too (probably in a separate PR) to be more consistent?

Reply:
@javanna I was thinking of leaving it as it is in this PR and maybe open a separate one to move all duplicated conditionals to a single place - wdyt?

Reply:
sounds good